File: backend/internal/controller/calculator.go
package controller

import (
	"errors"
	"net/http"

	"github.com/brunomasunaga/fs-calculator/backend/internal/dto"
	"github.com/brunomasunaga/fs-calculator/backend/internal/service"
	"github.com/gin-gonic/gin"
	"github.com/gin-gonic/gin/binding"
)

type CalculatorController struct {
	svc service.CalculatorService
}

func NewCalculatorController(svc service.CalculatorService) *CalculatorController {
	return &CalculatorController{svc: svc}
}

// Add calculates the sum of two operands.
// @Tags Operations
// @Summary Add two numbers
// @Accept json
// @Produce json
// @Param request body dto.BinaryOperationRequest true "Operands"
// @Success 200 {object} dto.CalculateResponse
// @Failure 400 {object} dto.ErrorResponse
// @Router /v1/operations/add [post]
func (cc *CalculatorController) Add(c *gin.Context) {
	req, ok := bindBinaryRequest(c)
	if !ok {
		return
	}

	c.JSON(http.StatusOK, dto.CalculateResponse{Result: cc.svc.Add(req.OperandA, req.OperandB)})
}

// Subtract calculates the difference between two operands.
// @Tags Operations
// @Summary Subtract two numbers
// @Accept json
// @Produce json
// @Param request body dto.BinaryOperationRequest true "Operands"
// @Success 200 {object} dto.CalculateResponse
// @Failure 400 {object} dto.ErrorResponse
// @Router /v1/operations/subtract [post]
func (cc *CalculatorController) Subtract(c *gin.Context) {
	req, ok := bindBinaryRequest(c)
	if !ok {
		return
	}

	c.JSON(http.StatusOK, dto.CalculateResponse{Result: cc.svc.Subtract(req.OperandA, req.OperandB)})
}

// Multiply calculates the product of two operands.
// @Tags Operations
// @Summary Multiply two numbers
// @Accept json
// @Produce json
// @Param request body dto.BinaryOperationRequest true "Operands"
// @Success 200 {object} dto.CalculateResponse
// @Failure 400 {object} dto.ErrorResponse
// @Router /v1/operations/multiply [post]
func (cc *CalculatorController) Multiply(c *gin.Context) {
	req, ok := bindBinaryRequest(c)
	if !ok {
		return
	}

	c.JSON(http.StatusOK, dto.CalculateResponse{Result: cc.svc.Multiply(req.OperandA, req.OperandB)})
}

// Divide calculates the quotient of two operands.
// @Tags Operations
// @Summary Divide two numbers
// @Accept json
// @Produce json
// @Param request body dto.BinaryOperationRequest true "Operands"
// @Success 200 {object} dto.CalculateResponse
// @Failure 400 {object} dto.ErrorResponse
// @Router /v1/operations/divide [post]
func (cc *CalculatorController) Divide(c *gin.Context) {
	req, ok := bindBinaryRequest(c)
	if !ok {
		return
	}

	result, err := cc.svc.Divide(req.OperandA, req.OperandB)
	if err != nil {
		writeError(c, err)
		return
	}

	c.JSON(http.StatusOK, dto.CalculateResponse{Result: result})
}

// Power raises the first operand to the power of the second operand.
// @Tags Operations
// @Summary Raise a number to a power
// @Accept json
// @Produce json
// @Param request body dto.BinaryOperationRequest true "Operands"
// @Success 200 {object} dto.CalculateResponse
// @Failure 400 {object} dto.ErrorResponse
// @Router /v1/operations/power [post]
func (cc *CalculatorController) Power(c *gin.Context) {
	req, ok := bindBinaryRequest(c)
	if !ok {
		return
	}

	c.JSON(http.StatusOK, dto.CalculateResponse{Result: cc.svc.Power(req.OperandA, req.OperandB)})
}

// Sqrt calculates the square root of a single operand.
// @Tags Operations
// @Summary Calculate a square root
// @Accept json
// @Produce json
// @Param request body dto.UnaryOperationRequest true "Operand"
// @Success 200 {object} dto.CalculateResponse
// @Failure 400 {object} dto.ErrorResponse
// @Router /v1/operations/sqrt [post]
func (cc *CalculatorController) Sqrt(c *gin.Context) {
	req, ok := bindUnaryRequest(c)
	if !ok {
		return
	}

	result, err := cc.svc.Sqrt(req.Operand)
	if err != nil {
		writeError(c, err)
		return
	}

	c.JSON(http.StatusOK, dto.CalculateResponse{Result: result})
}

// Percentage converts an operand into its percentage value.
// @Tags Operations
// @Summary Calculate percentage
// @Accept json
// @Produce json
// @Param request body dto.UnaryOperationRequest true "Operand"
// @Success 200 {object} dto.CalculateResponse
// @Failure 400 {object} dto.ErrorResponse
// @Router /v1/operations/percentage [post]
func (cc *CalculatorController) Percentage(c *gin.Context) {
	req, ok := bindUnaryRequest(c)
	if !ok {
		return
	}

	c.JSON(http.StatusOK, dto.CalculateResponse{Result: cc.svc.Percentage(req.Operand)})
}

func bindBinaryRequest(c *gin.Context) (dto.BinaryOperationRequest, bool) {
	var payload map[string]any
	var req dto.BinaryOperationRequest

	if err := c.ShouldBindBodyWith(&payload, binding.JSON); err != nil {
		writeError(c, err)
		return dto.BinaryOperationRequest{}, false
	}

	if err := c.ShouldBindBodyWith(&req, binding.JSON); err != nil {
		writeError(c, err)
		return dto.BinaryOperationRequest{}, false
	}

	if _, ok := payload["operand_a"]; !ok {
		writeError(c, errors.New("operand_a is required"))
		return dto.BinaryOperationRequest{}, false
	}

	if _, ok := payload["operand_b"]; !ok {
		writeError(c, errors.New("operand_b is required"))
		return dto.BinaryOperationRequest{}, false
	}

	return req, true
}

func bindUnaryRequest(c *gin.Context) (dto.UnaryOperationRequest, bool) {
	var payload map[string]any
	var req dto.UnaryOperationRequest

	if err := c.ShouldBindBodyWith(&payload, binding.JSON); err != nil {
		writeError(c, err)
		return dto.UnaryOperationRequest{}, false
	}

	if err := c.ShouldBindBodyWith(&req, binding.JSON); err != nil {
		writeError(c, err)
		return dto.UnaryOperationRequest{}, false
	}

	if _, ok := payload["operand"]; !ok {
		writeError(c, errors.New("operand is required"))
		return dto.UnaryOperationRequest{}, false
	}

	return req, true
}

func writeError(c *gin.Context, err error) {
	c.JSON(http.StatusBadRequest, dto.ErrorResponse{Error: err.Error()})
}

File: frontend/src/api/calculator.ts
import axios from 'axios'

import client from '@/api/client'

export interface BinaryOperationRequest {
  operand_a: number
  operand_b: number
}

export interface UnaryOperationRequest {
  operand: number
}

export interface CalculateResponse {
  result: number
}

export interface ErrorResponse {
  error: string
}

export type CalculatorOperation =
  | 'add'
  | 'subtract'
  | 'multiply'
  | 'divide'
  | 'power'
  | 'sqrt'
  | 'percentage'

export async function add(a: number, b: number): Promise<number> {
  return postBinary('/v1/operations/add', { operand_a: a, operand_b: b })
}

export async function subtract(a: number, b: number): Promise<number> {
  return postBinary('/v1/operations/subtract', { operand_a: a, operand_b: b })
}

export async function multiply(a: number, b: number): Promise<number> {
  return postBinary('/v1/operations/multiply', { operand_a: a, operand_b: b })
}

export async function divide(a: number, b: number): Promise<number> {
  return postBinary('/v1/operations/divide', { operand_a: a, operand_b: b })
}

export async function power(a: number, b: number): Promise<number> {
  return postBinary('/v1/operations/power', { operand_a: a, operand_b: b })
}

export async function sqrt(a: number): Promise<number> {
  return postUnary('/v1/operations/sqrt', { operand: a })
}

export async function percentage(a: number): Promise<number> {
  return postUnary('/v1/operations/percentage', { operand: a })
}

export async function calculate(
  operation: CalculatorOperation,
  a: number,
  b?: number,
): Promise<number> {
  switch (operation) {
    case 'add':
      return add(a, b ?? 0)
    case 'subtract':
      return subtract(a, b ?? 0)
    case 'multiply':
      return multiply(a, b ?? 0)
    case 'divide':
      return divide(a, b ?? 0)
    case 'power':
      return power(a, b ?? 0)
    case 'sqrt':
      return sqrt(a)
    case 'percentage':
      return percentage(a)
  }
}

async function postBinary(
  path: string,
  payload: BinaryOperationRequest,
): Promise<number> {
  try {
    const response = await client.post<CalculateResponse>(path, payload)
    return response.data.result
  } catch (error) {
    throw normalizeApiError(error)
  }
}

async function postUnary(
  path: string,
  payload: UnaryOperationRequest,
): Promise<number> {
  try {
    const response = await client.post<CalculateResponse>(path, payload)
    return response.data.result
  } catch (error) {
    throw normalizeApiError(error)
  }
}

function normalizeApiError(error: unknown): Error {
  if (axios.isAxiosError<ErrorResponse>(error)) {
    return new Error(error.response?.data?.error ?? error.message)
  }

  if (error instanceof Error) {
    return error
  }

  return new Error('Unexpected error')
}
